src/dbnode/storage/database.go

```diff
@@ -889,40 +895,130 @@ func (d *db) ReadEncoded(
 			xopentracing.Time("end", end),
 		)
 	}
+
 	defer sp.Finish()
+	return n.ReadEncoded(ctx, id, start, end)
+}
 
+func (d *db) WideQuery(
+	ctx context.Context,
+	namespace ident.ID,
+	query index.Query,
+	start time.Time,
+	shards []uint32,
+	iterOpts index.IterationOptions,
+) ([]ident.IndexChecksum, error) { // FIXME: change when exact type known.
 	n, err := d.namespaceFor(namespace)
 	if err != nil {
 		d.metrics.unknownNamespaceRead.Inc(1)
 		return nil, err
 	}
 
-	return n.ReadEncoded(ctx, id, start, end)
+	batchSize := d.opts.WideBatchSize()
+	doneCh := make(chan struct{})
+	collector := make(chan *ident.IDBatch)
+	blockSize := n.Options().IndexOptions().BlockSize()
+	opts := index.NewWideQueryOptions(start, batchSize, collector, blockSize, shards, iterOpts)
+	start, end := opts.StartInclusive, opts.EndExclusive
+	ctx, sp, sampled := ctx.StartSampledTraceSpan(tracepoint.DBWideQuery)
+	if sampled {
+		sp.LogFields(
+			opentracinglog.String("wideQuery", query.String()),
+			opentracinglog.String("namespace", namespace.String()),
+			opentracinglog.Int("batchSize", batchSize),
+			xopentracing.Time("start", start),
+			xopentracing.Time("end", end),
+		)
+	}
+
+	defer sp.Finish()
+
+	var (
+		c            ident.IndexChecksum
+		collected    = make([]ident.IndexChecksum, 0, 10)
+		collectorErr error
+	)
+
+	// Setup consumer
+	go func() {
+		for batch := range collector {
+			if collectorErr != nil {
+				batch.Done()
+				continue
+			}
+
+			for i, id := range batch.IDs {
+				useID := i == len(batch.IDs)-1
+				c, collectorErr = d.indexChecksum(ctx, namespace, n, id, start, useID)
+				if collectorErr == nil {
```

Comment:
Do we need to do anything if the error is not nil? Maybe break? Otherwise, collectorErr could get clobbered by successful following calls to `d.indexChecksum`

Reply:
Good call; wasn't too careful with this code since it's intended to be fairly temporary :+1: